### PR TITLE
Implement a CheckpointManager to save and load the checkpoint.

### LINKED ISSUE #771: Implement a CheckpointManager to save and load checkpoint.

### PR DESCRIPTION
**Pull request by workingloong**
### What changes were proposed in this pull request?

Implement a CheckpointManager to save and load the checkpoint.

### Why are the changes needed?

It is APIs for users to save and load checkpoint of DDP or FDSP model.

Fix #771 

### Does this PR introduce any user-facing change?

Yes. Users can use CheckpointManager to save and load checkpoint. For example.

``` python
>>> ckpt_manager = CheckpointManager(
>>>    model, optimizer, train_dataloader, "/tmp/checkpoint/"
>>> )
>>> ckpt_manager.save(0, 10)
>>> ckpt_manger.load()
```

### How was this patch tested?

We have tested it with a MNIST job.

### CODE REVIEW

**Comment by Antlera** [comment]
I will check this on nanogpt example.

**Comment by workingloong** [comment]
> I will check this on nanogpt example.

I have tested it on a node with 8 A100 GPUs and the logs are:

```


iter 97: loss 2.3601, time 19.84ms, mfu 0.74%, cuda memory 0.442G, lr 6.00e-04, total time 5.67s
--
iter 99: loss 2.3415, time 20.07ms, mfu 0.74%, cuda memory 0.442G, lr 6.00e-04, total time 5.71s
[2023-10-23 08:29:07,868] [INFO] [checkpoint.py:82:log_rank0] Save checkpoint of step=100 of epoch=0.
iter 100: loss 2.3628, time 22.38ms, mfu 0.73%, cuda memory 0.455G, lr 6.00e-04, total time 5.74s



iter 188: loss 1.9265, time 39.08ms, mfu 0.71%, cuda memory 0.455G, lr 6.00e-04, total time 7.52s
--
--> entering save model state
--> saving model ...
--> attempting to save model prefix -0.pt
--> saving as model name nanogpt-2023-10-23-08-29-09-0.pt


```

File: dlrover/trainer/torch/elastic/checkpoint.py [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+
+        Args:
+            epoch (int): the epoch index.
+            step (int): the iteration step in the epoch.
+        """
+        self.log_rank0(f"Save checkpoint of step={step} of epoch={epoch}.")
+        step = step + epoch * len(self.dataloader)
+        is_fsdp_model = isinstance(self.model, FSDP)
+        msd, osd = get_model_optim_state(
+            self.model, self.optimizer, is_fsdp_model
```
Comment (hxdtest) [inline dlrover/trainer/torch/elastic/checkpoint.py:104]: What if the model is wrapped by `DeepSpeedEngine`?

Comment (workingloong) [reply dlrover/trainer/torch/elastic/checkpoint.py]: I have refactored the implementation and developed independent classes for local, DDP and FSDP. If the model is wrapped by `DeepSpeedEngine`, we will raise a `NotImplementedError`. Later, we can implement the `CheckpointManager` for `DeepSpeedEngine`.

**Review by hxdtest — APPROVED** [review]

**Pull request merged**